compiler/z/codegen/ControlFlowEvaluator.cpp

```diff
@@ -338,21 +338,57 @@ xmaxxminHelper(TR::Node* node, TR::CodeGenerator* cg, TR::InstOpCode::Mnemonic c
    generateS390LabelInstruction(cg, TR::InstOpCode::label, node, cFlowRegionStart);
    cFlowRegionStart->setStartInternalControlFlow();
 
+   TR::RegisterDependencyConditions* deps = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(0, 3, cg);
+   deps->addPostConditionIfNotAlreadyInserted(lhsReg, TR::RealRegister::AssignAny);
+   deps->addPostConditionIfNotAlreadyInserted(rhsReg, TR::RealRegister::AssignAny);
+
+   //Checking common Condition Code and branching to cFlowRegionEnd
    generateRREInstruction(cg, compareRROp, node, lhsReg, rhsReg);
    generateS390BranchInstruction(cg, TR::InstOpCode::BRC, branchCond, node, cFlowRegionEnd);
 
-   generateRREInstruction(cg, moveRROp, node, lhsReg, rhsReg);
-
-   TR::RegisterDependencyConditions* deps = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(0, 2, cg);
+   //Label Symbols for handling possible NaN operand case for float/double
+   TR::LabelSymbol* nanLabel = NULL;
+   TR::LabelSymbol* notNaNLabel = generateLabelSymbol(cg);
+   //For Double & Float operands only, branching to nanLabel for Condition Code 3 when operands are unordered
+   if (node->getOpCode().isDouble() || node->getOpCode().isFloat())
+      {
+      nanLabel = generateLabelSymbol(cg);
+      generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_CC3, node, nanLabel);
+      //Branch to notNaNLabel as fallthrough
+      generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, notNaNLabel);
+
+      //Handle NaN case
+      generateS390LabelInstruction(cg, TR::InstOpCode::label, node, nanLabel);
+      //Allocate GPR for holding immediate NaN value for float/double
+      TR::Register * nanReg = cg->allocateRegister(TR_GPR);
+      //Handle NaN case for doubles
+      if (node->getOpCode().isDouble())
+         {
+         //Load immediate DOUBLE_NAN value into GPR, then move it to result FPR
+         genLoadLongConstant(cg, node, DOUBLE_NAN, nanReg, NULL, deps, NULL);
+         generateRRInstruction(cg, TR::InstOpCode::LDGR, node, lhsReg, nanReg);
+         }
+      else if (node->getOpCode().isFloat())
+         {
+         //Handle NaN case for floats by loading immediate FLOAT_NAN value into GPR
+         //Shift left the 32 least significant bits for preserving float representation before moving to FPR
+         generateRILInstruction(cg, TR::InstOpCode::LGFI, node, nanReg, FLOAT_NAN, NULL);
+         generateRSInstruction(cg, TR::InstOpCode::SLLG, node, nanReg, 32);
+         generateRRInstruction(cg, TR::InstOpCode::LDGR, node, lhsReg, nanReg);
+         }
+      deps->addPostConditionIfNotAlreadyInserted(nanReg, TR::RealRegister::AssignAny);
+      cg->stopUsingRegister(nanReg);
```

Comment:
Can you move this line to end of evaluator after you are attaching dependency to and of ICF label  (You just added the register to dep)